src/core/mempool.rs

```diff
@@ -750,6 +756,113 @@ impl MemPoolTxInfo {
     }
 }
 
+/// Used to locally cache nonces to avoid repeatedly looking them up in the nonce.
+struct NonceCache {
+    cache: HashMap<StacksAddress, u64>,
+    size: usize,
+}
+
+impl NonceCache {
+    const MAX_SIZE: usize = 1024 * 1024;
+
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+            size: 0,
+        }
+    }
+
+    fn get<C>(&mut self, address: &StacksAddress, clarity_tx: &mut C) -> u64
+    where
+        C: ClarityConnection,
+    {
+        match self.cache.get(address) {
+            Some(nonce) => *nonce,
+            None => {
+                let nonce = StacksChainState::get_nonce(clarity_tx, &address.clone().into());
+                // Simple size cap to the cache -- once it's full, all nonces
+                // will be looked up every time. This is bad for performance
+                // but is unlikely to occur due to the typical number of
+                // transactions processed before filling a block.
+                if self.size < Self::MAX_SIZE {
+                    self.cache.insert(address.clone(), nonce);
+                    self.size += 1;
+                }
+                nonce
+            }
+        }
+    }
+
+    fn increment(&mut self, address: StacksAddress) {
+        let nonce = self.cache.entry(address).or_insert(0);
+        *nonce += 1;
+    }
+}
+
+struct CandidateCache {
+    cache: VecDeque<MemPoolTxInfoPartial>,
+    next: VecDeque<MemPoolTxInfoPartial>,
+    size: usize,
+}
+
+impl CandidateCache {
+    const MAX_SIZE: usize = 64 * 1024;
+
+    fn new() -> Self {
+        Self {
+            cache: VecDeque::new(),
+            next: VecDeque::new(),
+            size: 0,
+        }
+    }
+
+    fn next(&mut self) -> Option<MemPoolTxInfoPartial> {
+        self.cache.pop_front()
+    }
+
+    fn push(&mut self, tx: MemPoolTxInfoPartial) {
+        if self.size < Self::MAX_SIZE {
+            self.next.push_back(tx);
+            self.size += 1;
+        }
+    }
+
+    fn reset(&mut self) {
```

Comment:
Could this cause `self.next` to grow unbound?  We aren't checking its length here, or in `Self::push`.

Reply:
It seems that the size of `self.next` can grow to `O(size of mempool)`, given that this method gets called once per iterator pass.  Is there a way to prevent this?

Reply:
Ah, I see my logic was off here -- size shouldn't have been reset to `0`. I'll just check the size of `self.next`. Also, I checked and saw that `VecDeque` does have an efficient `len()` method, so I'll just use that and remove the `size` field.

Reply:
@obycode I'm also saying that `.reset()` should check the size of `self.next`.  The `.append()` function will add every item in `self.cache` to `self.next`, which could push `self.next` to be well beyond `Self::MAX_SIZE`.  I think this is something we'd want to address in `.reset()`.

Reply:
Comment added in 27ee2f88c